crates/sui-network/src/discovery/mod.rs

```diff
@@ -0,0 +1,448 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anemo::{types::PeerEvent, Network, Peer, PeerId, Request, Response};
+use futures::StreamExt;
+use multiaddr::Multiaddr;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+use sui_config::p2p::{P2pConfig, SeedPeer};
+use tap::{Pipe, TapFallible};
+use tokio::sync::oneshot;
+use tokio::task::{AbortHandle, JoinSet};
+use tracing::{debug, info, trace};
+
+const TIMEOUT: Duration = Duration::from_secs(1);
+const MAX_NUMBER_OF_CONNECTIONS: usize = 20;
+const DISCOVERY_INTERVAL_PERIOD: Duration = Duration::from_millis(5_000);
+const ONE_DAY_MILLISECONDS: u64 = 24 * 60 * 60 * 1000;
+
+mod generated {
+    include!(concat!(env!("OUT_DIR"), "/sui.Discovery.rs"));
+}
+mod builder;
+mod server;
+#[cfg(test)]
+mod tests;
+
+pub use builder::{Builder, Handle, UnstartedDiscovery};
+pub use generated::{
+    discovery_client::DiscoveryClient,
+    discovery_server::{Discovery, DiscoveryServer},
+};
+pub use server::GetKnownPeersResponse;
+
+struct State {
+    our_info: Option<NodeInfo>,
+    connected_peers: HashMap<PeerId, ()>,
+    known_peers: HashMap<PeerId, NodeInfo>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub enum NodeType {
+    Validator,
```

Comment:
Another consistency-related question. I think I've unwittingly started a little battle in narwhal to stop calling things that are the same by multiple names. For ex:
digest, id -> digest
batch, payload -> batch
block, certificate -> certificate

Wondering if a similar concern applies here to "validator" vs. "authority"? I know the term "authority" appears a lot in Sui code. Are those two the same thing? If so, can we all agree on which of those two names we want to call it and at least start being standard in new code? Curious what you think on this. I know it's not the "fault" of this PR or anything :)

Reply:
So the term we've decided to use in our public facing docs is "Validator". Authority is a historical artifact that we haven't gone about renaming yet

Reply:
good to know thanks!